Lambdas/GHHooks/GHLazyMiddleware.swift

```diff
@@ -0,0 +1,113 @@
+import OpenAPIRuntime
+import Logging
+import Foundation
+import struct DiscordModels.Secret
+
+/// Adds some headers to all requests.
+/// Loads the GH token lazily to avoid additional secrets-manager costs.
+actor GHLazyMiddleware: ClientMiddleware {
+
+    /// Use the Secret type from DiscordBM to make sure that the token
+    /// doesn't end up in logs just because of a print()/log().
+    private var githubToken: Secret?
+    private let secretsRetriever: SecretsRetriever
+    private let logger: Logger
+
+    private var idGenerator = 0
+
+    private var isLoading = false
+    private var loadWaiters: [CheckedContinuation<Void, Never>] = []
+
+    init(secretsRetriever: SecretsRetriever, logger: Logger) {
+        self.secretsRetriever = secretsRetriever
+        self.logger = logger
+    }
+
+    func intercept(
+        _ request: Request,
+        baseURL: URL,
+        operationID: String,
+        next: @Sendable (Request, URL) async throws -> Response
+    ) async throws -> Response {
+        try await loadTokenIfNotLoaded()
+
+        var request = request
+        request.headerFields.addOrReplace(
+            name: "Accept",
+            value: "application/vnd.github.raw+json"
+        )
+        request.headerFields.addOrReplace(
+            name: "Authorization",
+            /// Token loaded so can force-unwrap.
+            value: "Bearer \(githubToken!.value)"
+        )
+        request.headerFields.addOrReplace(
+            name: "X-GitHub-Api-Version",
+            value: "2022-11-28"
+        )
+
+        self.idGenerator += 1
+        let requestID = idGenerator
+
+        logger.debug("Will send request to Github", metadata: [
+            "request": "\(request)",
+            "baseURL": .stringConvertible(baseURL),
+            "operationID": .string(operationID),
+            "requestID": .stringConvertible(requestID),
+        ])
+
+        let response = try await next(request, baseURL)
+
+        logger.debug("Got response from Github", metadata: [
+            "response": "\(response.fullDescription)",
+            "requestID": .stringConvertible(requestID),
+        ])
+
+        return response
+    }
+
+    /// Claims a loading lock then loads the token.
+    private func loadTokenIfNotLoaded() async throws {
+        if isLoading {
+            await withCheckedContinuation { continuation in
+                loadWaiters.append(continuation)
+            }
+        }
+
+        guard githubToken == nil else { return }
+
+        isLoading = true
+        defer { isLoading = false }
+
+        let githubToken = try await secretsRetriever.getSecret(arnEnvVarKey: "GH_TOKEN_ARN")
+        self.githubToken = Secret(githubToken)
+
+        for waiter in self.loadWaiters {
+            waiter.resume()
+        }
+        self.loadWaiters.removeAll()
+    }
```

Comment:
The reason i went through the trouble of a "manual lock" was that i had heard somethings about different threads not working well with a lock .... i'm still not sure but i guess an actor works on just 1 thread anyway so we should be good?!

I think the manual lock i did would work fine too (the locking mechanism isn't even needed yet based on the fact that the whole lambda's execution is sequential, but i wanted to be safe for the future)

I'll move the locking mechanism to the `SecretsRetriever` because i have more plans for it ... I could make the `discordClient` be created lazily ... it matters because `discordClient` needs the bot token secret and getting each secret costs $$$.

Reply:
ah so i didn't read the change originally ...

I don't think the locking mechanism does anything special ... i think the simpler locking mechanism i had is safe (i've tried in a lot of places, other than asking people about it .... I'm still not 100% sure it's safe, but i think it is)

Reply:
```swift
await withCheckedContinuation { continuation in
    self.stateLock.withLockVoid { // prevent concurrent mutation
        self.loadWaiters.append(continuation)
    }
}
```
specifically here, i don't think any "concurrent mutations" can happen. 
I know there is a `await` there which means there is a suspension point, but i think the suspension point is triggered after the whole closure is executed, and is only used to wait for resumption of the continuation. A month or two ago I looked into the implementation of it but not through the right place, so didn't come to a proper conclusion.

Reply:
```swift
self.stateLock.withLockVoid {
    self.isLoading = false
    while let waiter = self.loadWaiters.popLast() { // empties the array
        waiter.resume()
    }
    self.githubToken = try loadResult.get()
}
```
Here also ... there is no suspension point anyway so i don't think the lock does anything again.

Reply:
also the problem with `popLast()` is that it pops the last one  ... not the first one. There is also no `popFirst()`.
Not a big deal but i rather keep the order.

Reply:
As we discovered on Discord, this entire pile of nonsense I started was an utterly invalid thoroughly humiliating whopping error on my part 😅 That being said, just for the point of record - the waiter list by definition does not have a well-defined order, and even if it did, you wouldn't want it to. No matter how you consume it, by unblocking waiters in _any_ given order trashes the priority management semantics inherent to tasks and the lower-level locks they rely on (though moreso on macOS than Linux by a tremendous margin - Linux doesn't have priority-donating unfair locks, at least not in the sense of `os_unfair_lock_t`'s semantics). It's not even remotely worth caring about here, much like anything else I've said in this conversation thread 🤣, but it does mean that even if you were needing to deal with locking, the ordering of the waiters can never be well-defined without explicit serialization (which defeats the point) and resuming them in _any_ given order is always arguably wrong anyway, so random or inverse is as good as anything else. (Countdown until I figure out that this explanation is also irrelevant or based on a fundamental misreading: 10... 5... 8... -∞... π... TREE(TREE(TREE(pow(2, 128)))), √-1... 😂)